Src/PerformanceCollector/Shared/Implementation/QuickPulse/QuickPulseServiceClient.cs

```diff
@@ -0,0 +1,228 @@
+
+        private const int RetryCount = 3;
+        
+        private readonly string instanceName;
+
+        private readonly string version;
+
+        private readonly TimeSpan timeout = TimeSpan.FromSeconds(3);
+        
+        public QuickPulseServiceClient(Uri serviceUri, string instanceName, string version, TimeSpan? timeout = null)
+        {
+            this.ServiceUri = serviceUri;
+            this.instanceName = instanceName;
+            this.version = version;
+            this.timeout = timeout ?? this.timeout;
+        }
+
+        public Uri ServiceUri { get; }
+
+        public bool? Ping(string instrumentationKey, DateTime timestamp)
+        {
+            MemoryStream bodyStream = this.WritePingData(timestamp);
+
+            var path = string.Format(CultureInfo.InvariantCulture, "ping?ikey={0}", instrumentationKey);
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, true);
+            if (response == null)
+            {
+                return null;
+            }
+            
+            return ProcessResponse(response);
+        }
+        
+        public bool? SubmitSamples(IEnumerable<QuickPulseDataSample> samples, string instrumentationKey)
+        {
+            MemoryStream bodyStream = WriteSamples(samples, instrumentationKey);
+            
+            var path = string.Format(CultureInfo.InvariantCulture, "post?ikey={0}", HttpUtility.UrlEncode(instrumentationKey));
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, false);
+            if (response == null)
+            {
+                return null;
+            }
+
+            return ProcessResponse(response);
+        }
+
+        private static bool? ProcessResponse(HttpWebResponse response)
+        {
+            bool isSubscribed;
+            if (!bool.TryParse(response.GetResponseHeader(XMsQpsSubscribedHeaderName), out isSubscribed))
+            {
+                return null;
+            }
+
+            return isSubscribed;
+        }
+
+        private static double Round(double value)
+        {
+            return Math.Round(value, 4, MidpointRounding.AwayFromZero);
+        }
+
+        private MemoryStream WritePingData(DateTime timestamp)
+        {
+            var bodyStream = new MemoryStream();
+
+            var dataPoint = new MonitoringDataPoint
+            {
+                Version = this.version,
+                //InstrumentationKey = instrumentationKey,
+                Instance = this.instanceName,
+                Timestamp = timestamp
+            };
+
+            var serializer = new DataContractJsonSerializer(typeof(MonitoringDataPoint));
+            serializer.WriteObject(bodyStream, dataPoint);
+            bodyStream.Position = 0;
+
+            return bodyStream;
+        }
+
+        private MemoryStream WriteSamples(IEnumerable<QuickPulseDataSample> samples, string instrumentationKey)
```

Comment:
I wonder why you taking `instrumentationKey` as an argument and `instanceName` only populated once. Shouldn't you take if from telemetry items all the time?